src/core/net/ip6.cpp

```diff
@@ -977,6 +981,12 @@ Error Ip6::PassToHost(Message           &aMessage,
     // Do not pass IPv6 packets that exceed kMinimalMtu.
     VerifyOrExit(aMessage.GetLength() <= kMinimalMtu, error = kErrorDrop);
 
+#if OPENTHREAD_CONFIG_IP6_CORE_FIREWALL_ENABLE
+    VerifyOrExit(!Get<Mle::Mle>().IsMeshLocalAddress(aMessageInfo.GetSockAddr()) ||
+                     otPlatHostHasAddress(&aMessageInfo.GetPeerAddr()) || aMessageInfo.GetPeerAddr().IsMulticast(),
```

Comment:
It seems there is no guarantee that `otPlatHostHasAddress` returns true for an address of this Thread interface.

What about checking the `receive` flag and always allow packets when it's `true`?

Reply:
nit: 
switch the order of `otPlatHostHasAddress(&aMessageInfo.GetPeerAddr())` and `aMessageInfo.GetPeerAddr().IsMulticast()` to try to avoid the overhead of `otPlatHostHasAddress`